File: Alchem/core/artifact/artifact.ts
// Artifact — Object Data. Plan de construction, pas du code.

import type { TemplateId } from "./template";

export type ArtifactStatus = "prototype" | "in_progress" | "working" | "finished";
export type ArtifactType   = "code" | "documentation" | "module" | "config";

export interface ArtifactHeader {
  readonly artifactId: number;
  readonly name:       string;
  readonly createdBy:  string;
  readonly timestamp:  number;        // Unix ms
  readonly status:     ArtifactStatus;
  readonly type:       ArtifactType;
  readonly frequency?: string;        // ex. "639"
  readonly upgraded?:  boolean;
  readonly parentId?:  number;        // si c'est une mise à niveau
}

export interface Artifact extends ArtifactHeader {
  readonly templates:  readonly TemplateId[];   // référence, pas le code
  readonly tickOrder:  readonly TemplateId[];   // Fixed Sacred Order
  readonly variables:  Readonly<Record<string, unknown>>;
}

File: src/artifact/artifact.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ArtifactStatus {
    Prototype,
    #[serde(rename = "in_progress")]
    InProgress,
    Working,
    Finished,
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ArtifactType {
    Code,
    Documentation,
    Module,
    Config,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Artifact {
    #[serde(rename = "artifactId")]
    pub artifact_id: u32,

    pub name:        String,

    #[serde(rename = "createdBy")]
    pub created_by:  String,

    pub timestamp:   u64,
    pub status:      ArtifactStatus,

    #[serde(rename = "type")]
    pub kind:        ArtifactType,

    pub frequency:   Option<String>,

    #[serde(rename = "tickOrder")]
    pub tick_order:  Vec<u32>,

    pub variables:   serde_json::Value,
}
